### ISSUE TITLE
vp->iocount>0 assertion fails in vnode_put on file copy

### ISSUE THREAD

**Issue by jheuking** [openzfsonwindows/ZFSin#55]
I tried to copy a few dummy .txt-files to the zpool one by one. After some files vnode_put is called with the callstack below and the vnode->iocount is already 0.
I would suggest this is a timing problem. I have the feeling, that the problem occurs earlier, when I copy the files faster one behind the other.

```
>	ZFSin!vnode_put+0xa7 [c:\users\julianh\source\repos\zfsin\zfsin\spl\module\spl\spl-vnode.c @ 767]	C/C++/ASM
 	ZFSin!spl_rele_async+0x25 [c:\users\julianh\source\repos\zfsin\zfsin\spl\module\spl\spl-vnode.c @ 586]	C/C++/ASM
 	ZFSin!taskq_thread+0x494 [c:\users\julianh\source\repos\zfsin\zfsin\spl\module\spl\spl-taskq.c @ 1610]	C/C++/ASM
 	nt!PspSystemThreadStartup+0x47	C/C++/ASM
 	nt!KiStartSystemThread+0x16	C/C++/ASM

```

**Comment by lundman** [comment]
`zfs_purgedir(znode_t *dzp)` looks a bit suspicious, could that create two VN_RELE_ASYNC() calls maybe

**Comment by jheuking** [comment]
It seems that VN_RELE_ASYNC is only called before a crash via zfs_get_data->zfs_get_done.

```
dispatcher: enter: major 9: minor 0: IRP_MJ_FLUSH_BUFFERS: type 0x6
  fsDispatcher: enter: major 9: minor 0: IRP_MJ_FLUSH_BUFFERS fsDeviceObject
flush_buffers: 
vnode_getwithref: get vnode \ iocount 1
zfs_fsync
zil_commit zp \
zil_commit
zil_commit_impl
zfs_get_data: zp \mnt\test - Copy (3).txt
zfs_get_done: zfs_range_unlock skipped

            vn_rele_async: start task rel vnode \mnt\test - Copy (3).txt iocount 0
[...]
            spl_rele_async: rel vnode \mnt\test - Copy (3).txt iocount 0
            vnode_put: release vnode \mnt\test - Copy (3).txt iocount 0

*** Assertion failed: vp->v_iocount > 0
```

In other cases, when IRP_MJ_FLUSH_BUFFERS is received, zfs_get_data doesn't get called. Therefore vn_rele_async doesn't get called either and everything is fine. Looks like this:

```
dispatcher: enter: major 9: minor 0: IRP_MJ_FLUSH_BUFFERS: type 0x6
  fsDispatcher: enter: major 9: minor 0: IRP_MJ_FLUSH_BUFFERS fsDeviceObject
flush_buffers: 
vnode_getwithref: get vnode \ iocount 1
zfs_fsync
zil_commit zp \
zil_commit
zil_commit_impl
dispatcher: enter: major 0: minor 0: IRP_MJ_CREATE: type 0x6
...
```

So the problem should be somewhere here, I guess..

**Comment by jheuking** [comment]
should be fixed with this, hopefully #57 😬
